flax/linen/linear.py

```diff
@@ -1009,6 +1011,8 @@ def __call__(self, inputs: Array) -> Array:
     (embedding,) = promote_dtype(
       self.embedding, dtype=self.dtype, inexact=False
     )
+    if (self.num_embeddings == 1):
+      return jnp.broadcast_to(embedding, inputs.shape + (self.features,))
```

Comment:
If we pass in index values in `inputs` that are greater or equal to `self.num_embeddings`, the current expected behavior is that it returns a `nan` embedding. 

With this change, we would still be able to get the embedding matrix (e.g. if we pass in `inputs` that contains index value 2), since the `if` condition only uses `inputs.shape` and not `inputs`. 

This could introduce silent errors if the user isn't aware of this special case with `self.num_embeddings==1`, since no `nan` will be returned if they try to index index past `self.num_embeddings`.

We should either add a check in this `if` block to check for over-indexing and if so, return `nan` like the previous behavior, or we can keep this change as in and describe this special case in the docstring.

What do you think @cgarciae?

Reply:
If we prefer `nan` in this case, It can be
```
    if (self.num_embeddings == 1):
      return jax.lax.cond((inputs==0).all(),
                          lambda x: jnp.broadcast_to(embedding,x.shape+(self.features,)),
                          lambda x: jnp.take(embedding, x, axis=0),
                          inputs)
```

Reply:
Sounds good to me! Can you also write a test in [`tests/linen/linen_linear_test.py`](https://github.com/google/flax/blob/main/tests/linen/linen_linear_test.py#L1046) and [`flax/experimental/nnx/tests/nn/test_linear.py`](https://github.com/google/flax/blob/main/flax/experimental/nnx/tests/nn/test_linear.py#L69) to confirm the `nan` behavior for both `self.num_embeddings == 1` and `self.num_embeddings > 1`

Reply:
Thank you. I have updated the code for the `nan` case and added two tests.

Reply:
Great, thanks! Could you also run `ruff format flax/experimental/nnx/nnx/nn/linear.py flax/experimental/nnx/tests/nn/test_embed.py flax/linen/linear.py tests/linen/linen_linear_test.py`

Reply:
Thanks! I have updated these files.

Reply:
@jianyizh I think you should use `jnp.where` instead of `cond` for performance, also because here you are conditioning on the entire batch being within bounds.

Reply:
I have updated to jnp.where. Actually, since the inputs are constants in the models when `num_embedding == 1`, there is no performance difference. XLA do some optimizations like constant folding, both `where` and `cond` are canceled. The two HLO graphs are both simplified to reduce_sum on backward.